syntax = "proto3";

package protobufs.services.v1;

import "google/protobuf/timestamp.proto";
import "protobufs/common/v1/solution.proto";
import "protobufs/common/v1/status.proto";
import "protobufs/common/v1/language.proto";

// A service to handle receiving code submissions
service SubmissionService {
    // Submit code for evaluation.
    rpc SubmitCode(SubmitCodeRequest) returns (SubmitCodeResponse) { }
    // Gets submission data.
    rpc GetSubmission(GetSubmissionRequest) returns (GetSubmissionResponse) { }
    // Get all submissions made by a user.
    rpc GetUserSubmissions(GetUserSubmissionsRequest) returns (GetUserSubmissionsResponse) { }
    // Get all submissions for a problem.
    rpc GetProblemSubmissions(GetProblemSubmissionsRequest) returns (GetProblemSubmissionsResponse) { }
    // Get progress of a submission
    rpc GetSubmissionProgress(GetSubmissionProgressRequest) returns (GetSubmissionProgressResponse) { }
    // Update progress of a submission
    rpc UpdateSubmissionProgress(UpdateSubmissionProgressRequest) returns (UpdateSubmissionProgressResponse) { }

    rpc GetSubmissionStateForUser(GetSubmissionStateForUserRequest) returns (GetSubmissionStateForUserResponse) { }
}

message SubmitCodeRequest {
    string problem_id = 1;
    string user_id = 2;
    common.v1.ProgrammingLanguage language = 3;
    repeated common.v1.SolutionFile files = 4;
}

message SubmitCodeResponse {
    string submission_id = 1;
}

message GetSubmissionRequest {
    string submission_id = 1;
}

message GetSubmissionResponse {
    string problem_id = 1;
    string user_id = 2;
    string submission_id = 3;
    common.v1.ProgrammingLanguage language = 4;
    repeated common.v1.SolutionFile files = 5;
    repeated SubmissionTestResult test_results = 6;
    google.protobuf.Timestamp submission_time = 7;
}

message GetUserSubmissionsRequest {
    string user_id = 1;
}

message GetUserSubmissionsResponse {
    repeated string submission_id = 1;
}

message GetProblemSubmissionsRequest {
    string problem_id = 1;
}

message GetProblemSubmissionsResponse {
    repeated string submission_id = 1;
}

message GetSubmissionProgressRequest {
    string problem_id = 1;
    string user_id = 2;
}

message GetSubmissionProgressResponse {
    bytes data = 1;
    google.protobuf.Timestamp last_updated = 2;
}

message UpdateSubmissionProgressRequest {
    string problem_id = 1;
    string user_id = 2;
    bytes data = 3;
}

message UpdateSubmissionProgressResponse {
    bool success = 1;
}

message SubmissionStatusEvent {
    common.v1.SubmissionStatus state = 1;
    SubmissionTestResult result = 2;
}

message SubmissionTestResult {
    string test_id = 1;
    bool passed = 2;
    string output = 3;
    float runtime = 4;
}

message GetSubmissionStateForUserRequest {
    string problem_id = 1;
    string user_id = 2;
}

message GetSubmissionStateForUserResponse {
    uint32 state = 1;
    uint32 test_passed = 3;
    uint32 test_total = 4;
}